sdk/core/azure-core/src/main/java/com/azure/core/credential/SimpleTokenCache.java

```diff
@@ -3,53 +3,135 @@
 
 package com.azure.core.credential;
 
+import com.azure.core.util.logging.ClientLogger;
 import reactor.core.publisher.FluxSink;
 import reactor.core.publisher.FluxSink.OverflowStrategy;
 import reactor.core.publisher.Mono;
 import reactor.core.publisher.ReplayProcessor;
 
+import java.time.Duration;
+import java.time.OffsetDateTime;
 import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Predicate;
 import java.util.function.Supplier;
 
 /**
  * A token cache that supports caching a token and refreshing it.
  */
 public class SimpleTokenCache {
-    private static final int REFRESH_TIMEOUT_SECONDS = 30;
-
     private final AtomicBoolean wip;
-    private AccessToken cache;
+    private volatile AccessToken cache;
+    private volatile OffsetDateTime nextTokenRefresh = OffsetDateTime.now();
     private final ReplayProcessor<AccessToken> emitterProcessor = ReplayProcessor.create(1);
     private final FluxSink<AccessToken> sink = emitterProcessor.sink(OverflowStrategy.BUFFER);
     private final Supplier<Mono<AccessToken>> tokenSupplier;
+    private final Predicate<AccessToken> shouldRefresh;
+    private final Duration refreshRetryTimeout;
+    private final ClientLogger logger = new ClientLogger(SimpleTokenCache.class);
 
     /**
      * Creates an instance of RefreshableTokenCredential with default scheme "Bearer".
      *
      * @param tokenSupplier a method to get a new token
      */
     public SimpleTokenCache(Supplier<Mono<AccessToken>> tokenSupplier) {
+        this(tokenSupplier, new TokenRefreshOptions());
+    }
+
+    /**
+     * Creates an instance of RefreshableTokenCredential with default scheme "Bearer".
+     *
+     * @param tokenSupplier a method to get a new token
+     * @param tokenRefreshOptions the options to configure the token refresh behavior
+     */
+    public SimpleTokenCache(Supplier<Mono<AccessToken>> tokenSupplier, TokenRefreshOptions tokenRefreshOptions) {
         this.wip = new AtomicBoolean(false);
         this.tokenSupplier = tokenSupplier;
+        this.shouldRefresh = accessToken -> OffsetDateTime.now().isAfter(accessToken.getExpiresAt()
+            .minus(tokenRefreshOptions.getTokenRefreshOffset()));
+        this.refreshRetryTimeout = tokenRefreshOptions.getTokenRefreshRetryTimeout();
     }
 
     /**
      * Asynchronously get a token from either the cache or replenish the cache with a new token.
      * @return a Publisher that emits an AccessToken
      */
     public Mono<AccessToken> getToken() {
-        if (cache != null && !cache.isExpired()) {
-            return Mono.just(cache);
-        }
-        return Mono.defer(() -> {
+        try {
             if (!wip.getAndSet(true)) {
```

Comment:
This is attempting to prevent multiple refresh operations from occurring at the same time correct?

Reply:
Yes